### PR TITLE
feat: default serialization for custom objects

### PR DESCRIPTION
**Pull request by iamogbz**
## Description

Adds a default serialization for custom objects

## Related Issues

N/A

## Checklist

- [x] This PR has sufficient test coverage.
- [x] I have updated the CHANGELOG.md.

## Additional Comments

Makes adoption easier

### CODE REVIEW

File: tests/examples/__snapshots__/test_custom_object_repr.ambr [hunk 1]
```diff
@@ -1,10 +1,18 @@
-# name: test_snapshot_attr_class
-  MyCustomReprClass(prop1=1, prop2='a', prop3={1, 2, 3})
+# name: test_snapshot_custom_class
+  <class 'MyCustomClass'> {
+  prop1=1
```
Comment (noahnu) [inline tests/examples/__snapshots__/test_custom_object_repr.ambr]: We should indent the props

File: src/syrupy/extensions/amber.py [hunk 2]
```diff
@@ -192,7 +192,26 @@ def serialize_iterable(
...
+                f"{serialized_key}={serialized_value.lstrip(cls._indent)}\n"
+                for serialized_key, serialized_value in (
+                    (
+                        cls.with_indent(name, depth=depth + 1),
+                        cls.serialize(
+                            data=getattr(data, name), depth=depth + 1, visited=visited
+                        ),
+                    )
+                    for name in cls.sort(dir(data))
+                    if not name.startswith("__")
```
Comment (noahnu) [inline src/syrupy/extensions/amber.py]: Thoughts on relaxing this to a single underscore?

Comment (iamogbz) [reply src/syrupy/extensions/amber.py]: Sure, it would prevent protected attributes from getting recorded

File: src/syrupy/extensions/amber.py [hunk 3]
```diff
@@ -188,6 +188,18 @@ def serialize_iterable(
...
+    def serialize_callable(
+        cls,
+        data: "SerializableData",
+        *,
+        depth: int = 0,
+        visited: Optional[Set[Any]] = None,
+    ) -> str:
+        return cls.with_indent(
+            f"{cls.object_type(data)} ({cls.MarkerDepthMax()})", depth
+        )
```
Comment (noahnu) [inline src/syrupy/extensions/amber.py]: There are some options here. Should include the function name? Doc string? type information if available?

What's a use case where someone would snapshot a function?

Comment (iamogbz) [reply src/syrupy/extensions/amber.py]: We can definitely get the type information, but I was wondering if it's better to just ignore methods when in a class. That way we won't need to account for callable.

Comment (noahnu) [reply src/syrupy/extensions/amber.py]: Personally I'm in favour of ignoring and leaving this up to a community built extension of needed, or open issues.

**Pull request merged**